### ISSUE TITLE
Arbitrary docker daemon fail

### ISSUE THREAD

**Issue by ese** [kubernetes/kops#2178]
I notice that sometimes I get docker daemon stucked in my nodes. kubelet is unable to launch new containers and all pods scheduled to that specific node stuck in Unknown status forever.
I couldn't find the root of the problem but it can be solved with a simple `service docker restart` so maybe it is possible to have a check in the docker runtime and restart it if something don't work and at the end restart the node if the problem is unrecoverable.

kubectl get pods

```
$kubectl get pods
w26-194366978-75sm1                       1/1       Unknown       0          1d        100.96.19.48    ip-172-21-97-105.eu-west-1.compute.internal
w26-4047883840-hdvgt                      0/1       Unknown       0          6h        <none>          ip-172-21-97-105.eu-west-1.compute.internal
```
```
$ kubectl describe node ip-172-21-97-105.eu-west-1.compute.internal
Name:			ip-172-21-97-105.eu-west-1.compute.internal
Role:			
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/instance-type=c4.2xlarge
			beta.kubernetes.io/os=linux
			failure-domain.beta.kubernetes.io/region=eu-west-1
			failure-domain.beta.kubernetes.io/zone=eu-west-1c
			kubernetes.io/hostname=ip-172-21-97-105.eu-west-1.compute.internal
			spot=true
Taints:			spot=mid:NoSchedule
CreationTimestamp:	Thu, 09 Mar 2017 17:29:19 +0100
Phase:			
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  OutOfDisk 		False 	Thu, 23 Mar 2017 21:54:11 +0100 	Thu, 23 Mar 2017 20:22:48 +0100 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Thu, 23 Mar 2017 21:54:11 +0100 	Thu, 09 Mar 2017 17:29:19 +0100 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  DiskPressure 		False 	Thu, 23 Mar 2017 21:54:11 +0100 	Thu, 09 Mar 2017 17:29:19 +0100 	KubeletHasNoDiskPressure 	kubelet has no disk pressure
  Ready 		False 	Thu, 23 Mar 2017 21:54:11 +0100 	Thu, 23 Mar 2017 20:22:48 +0100 	KubeletNotReady 		container runtime is down
  NetworkUnavailable 	False 	Thu, 09 Mar 2017 17:29:24 +0100 	Thu, 09 Mar 2017 17:29:24 +0100 	RouteCreated 			RouteController created a route
Addresses:		172.21.97.105,172.21.97.105,ip-172-21-97-105.eu-west-1.compute.internal
Capacity:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					8
 memory:				15403452Ki
 pods:					110
Allocatable:
 alpha.kubernetes.io/nvidia-gpu:	0
 cpu:					8
 memory:				15403452Ki
 pods:					110
System Info:
 Machine ID:			245e887d4e164daca68528a1e5f73337
 System UUID:			EC2A0CBA-B433-7AD4-E701-570BC679B13A
 Boot ID:			a463cb0e-2abe-467d-bf9f-ace2712098b7
 Kernel Version:		4.4.41-k8s
 OS Image:			Debian GNU/Linux 8 (jessie)
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.12.3
 Kubelet Version:		v1.5.3
 Kube-Proxy Version:		v1.5.3
PodCIDR:			100.96.19.0/24
ExternalID:			i-0ac3b16a045247881
Non-terminated Pods:		(25 in total)
[...]
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.
  CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ------------	----------	---------------	-------------
  4800m (60%)	3600m (45%)	5112Mi (33%)	3512Mi (23%)
Events:
  FirstSeen	LastSeen	Count	From							SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----							-------------	--------	------			-------
  1h		2m		30	{kubelet ip-172-21-97-105.eu-west-1.compute.internal}			Warning		ContainerGCFailed	operation timeout: context deadline exceeded
```

Last logs from `journalctl -u docker.service`
```
Mar 17 16:22:20 ip-172-21-97-105 dockerd[1079]: http: multiple response.WriteHeader calls
Mar 19 00:17:52 ip-172-21-97-105 dockerd[1079]: http: multiple response.WriteHeader calls
Mar 19 00:17:53 ip-172-21-97-105 dockerd[1079]: http: multiple response.WriteHeader calls
Mar 23 19:22:44 ip-172-21-97-105 dockerd[1079]: http: multiple response.WriteHeader calls
```

**Comment by ese** [comment]
GKE use this script
https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/health-monitor.sh

**Comment by andrewsykim** [comment]
@ese if I recall correctly, kops has a similar script https://github.com/kubernetes/kops/blob/master/upup/models/nodeup/docker/_systemd/_debian_family/files/opt/kubernetes/helpers/docker-healthcheck.

**Comment by ese** [comment]
@andrewsykim I don't see that script running in my nodes

**Comment by andrewsykim** [comment]
And you're using the default image?

**Comment by dmcnaught** [comment]
I'm seeing this too with default image, and the docker-healthcheck.service also returns no logs (K8s 1.5.2, kops 1.5.1)

**Comment by dmcnaught** [comment]
I think these tickets are related: kubernetes/kubernetes#42164
kubernetes/kubernetes#39028

**Comment by Bekt** [comment]
Sometimes (most of the time actually) it's impossible to restart/kill the docker daemon (the infamous kernel issue) -- the only way is to restart the node.

**Comment by dmcnaught** [comment]
I've been restarting docker everytime to fix the problem @Bekt 
```bash
root@ip-10-153-79-152:~# /etc/init.d/docker restart
[ ok ] Restarting docker (via systemctl): docker.service.
```

**Comment by andrewsykim** [comment]
haha @Bekt @dmcnaught I have felt the pain of this infamous kernel issue you talk about. I have actually worked around it by tailing my kernal logs and rebooting entire nodes to work around this :(

**Comment by andrewsykim** [comment]
Luckily with docker version > 1.11 this bug happens less often

**Comment by Bekt** [comment]
@dmcnaught https://github.com/kubernetes/kubernetes/issues/42164 happened to me again this morning, and I ssh-ed into the node to verify this. docker (1.12.6) is unresponsive, probably the same issue as https://github.com/coreos/bugs/issues/1654, https://github.com/docker/docker/issues/12606

```
core@ip-172-32-0-248 ~ $ docker --version
Docker version 1.12.6, build d5236f0
core@ip-172-32-0-248 ~ $ docker info
^C
# the command above just hangs

core@ip-172-32-0-248 ~ $ sudo systemctl restart docker
^C
# also hangs

core@ip-172-32-0-248 ~ $ systemctl status docker
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; disabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─40-flannel.conf
   Active: deactivating (stop-sigterm) since Mon 2017-04-03 15:19:08 UTC; 16min ago
```

[docker logs](https://pastebin.com/raw/tjvEh0hH)

I don't think this is a kops issue.

**Comment by andrewsykim** [comment]
This is definitely a linux kernel + docker issue. I would definitely recommend finding out why your kops distro doesn't have the docker health check script because I find that has helped me a lot.

**Comment by Bekt** [comment]
It helps in some cases, but it won't help at all when docker itself is unresponsive (see my previous comment snippet). I imagine this is due to some kernel deadlock.

**Comment by ese** [comment]
closed by #2299
